File: 8kyu/(8 kyu) MakeUpperCase/(8 kyu) MakeUpperCase.java
// 1 Simple solution
class Upper {
    public static String makeUpperCase(String str) {
        String result = "";

        for (int i = 0; i < str.length(); i++) {
            char cc = str.charAt(i);
            if (cc >= 'a' && cc <= 'z') {
                cc = (char) (cc - 32);
            }
            result += cc;
        }

        return result;
    }
}

// 2 Optimized solution
class Upper {
    public static String makeUpperCase(String str) {
        return str
            .chars()
            .mapToObj(c -> (char) c)
            .map(c -> (c >= 'a' && c <= 'z') ? (char) (c - 32) : c)
            .collect(StringBuilder::new, StringBuilder::append, StringBuilder::append)
            .toString();
    }
}

// 3 Clever solution
class Upper {
    public static String makeUpperCase(String str) {
        return str.toUpperCase();
    }
}

File: 8kyu/(8 kyu) MakeUpperCase/(8 kyu) MakeUpperCase.rs
// 1 Simple solution
fn make_upper_case(s: &str) -> String {
    let mut result = String::new();
    for c in s.chars() {
        if c >= 'a' && c <= 'z' {
            result.push(((c as u8) - 32) as char);
        } else {
            result.push(c);
        }
    }
    result
}

// 2 Optimized solution
fn make_upper_case(s: &str) -> String {
    s.chars()
        .map(|c| if c.is_ascii_lowercase() { ((c as u8) - 32) as char } else { c })
        .collect()
}

// 3 Clever solution
fn make_upper_case(s: &str) -> String {
    s.to_ascii_uppercase()
}

// 4 Coding golf
static make_upper_case: fn(&str) -> String = str::to_uppercase;
